cache/charged_cache.cc

```diff
@@ -0,0 +1,113 @@
+//  Copyright (c) Meta Platforms, Inc. and affiliates.
+//  This source code is licensed under both the GPLv2 (found in the
+//  COPYING file in the root directory) and Apache 2.0 License
+//  (found in the LICENSE.Apache file in the root directory).
+
+#include "cache/charged_cache.h"
+
+#include "cache/cache_reservation_manager.h"
+
+namespace ROCKSDB_NAMESPACE {
+
+ChargedCache::ChargedCache(std::shared_ptr<Cache> cache,
+                           std::shared_ptr<Cache> block_cache)
+    : cache_(cache),
+      cache_res_mgr_(std::make_shared<ConcurrentCacheReservationManager>(
+          std::make_shared<
+              CacheReservationManagerImpl<CacheEntryRole::kBlobCache>>(
+              block_cache))) {}
+
+Status ChargedCache::Insert(const Slice& key, void* value, size_t charge,
+                            DeleterFn deleter, Handle** handle,
+                            Priority priority) {
+  Status s = cache_->Insert(key, value, charge, deleter, handle, priority);
+  if (s.ok()) {
+    // Insert may cause the cache entry eviction if the cache is full. So we
+    // directly call the reservation manager to update the total memory used
+    // in the cache.
+    assert(cache_res_mgr_);
+    cache_res_mgr_->UpdateCacheReservation(cache_->GetUsage())
```

Comment:
Since we always call `EvictFromLRU(e->total_charge, &last_reference_list);` first in `LRUCacheShard::InsertItem()`, I thought Status::MemoryLimit() will only happen when the inserted entry size is larger than the block cache capacity? If this is the case, the insertion into the blob cache already failed (`Status s = cache_->Insert(key, value, helper, charge, handle, priority);`), because the capacity of the blob cache is less than or equal to the block cache.

Reply:
> We can make this change in a follow-up PR.

@ltamasi @gangliao sounds good to me!